### ISSUE TITLE
Potential extension to trace indirect contacts (n-th degree).

### ISSUE THREAD

**Issue by solso** [DP-3T/documents#160]
It should be possible to track n-degree contacts on a privacy preserving fashion. 

(This was written before I was aware of PEPP (for reference only https://gist.github.com/solso/d3d298b75eaa917b855773118f0872d6). I'll try to summarize the key-points adapted to the terminology of this project. I would advise to read the gist after the explanation below)

The gist of it is to be able to evaluate the risk of any contact, iteratively, so that risk assessment is propagated through the network. Direct contacts totally disregard any environmental consideration, which is important on fomite driven transmission. Multiple contact degrees are able to account (partially) for it. Think of a contaminated supermarket, through indirect contacts all people who have been in that supermarket might be alerted though the pivots on the contact traces (the clerks, cashiers, etc).

The proposal has an important assumption, basically, not fully decentralized. But before jumping to conclusions please continue reading :-)

It assumes a globally accessible whiteboard, which has only 2 operations: 1) set(EphID, E(kEphID, p_s)) and 2) get(truncated(EphID)).

The whiteboard contains the self-calculated risk profile (a number in the (0..1]) range for any EphID. 

Any client can do a get by EphID (truncated) and will receive the set of <EphID, E(p_s)> on anyone on that range. The client will only see the actual value of p_s if and only if it has seen the kEphID. This is meant to prevent scanning for EphID to get the value of p_s on people who you have not met.

The client calculates their p_s locally, fetching the data on any observed contacts for which he has EphId and kEphID. After risk is calculated, it will be set on the whiteboard for each active EphID in the time interval, basically, update you current risk across all the EphID used in the last 14 days. This process is done multiple times, iteratively, by all the population. So that the risk is propagated to the network until steady state. This should converge after few iterations. 

Needless to say, all communication between clients and whiteboard should go through anonymization layer like TOR. Of course, no bulk messages or pipeline. Temporal and spatial correlations must be resolved on the client-side.

The end result is that you can estimate your risk, based not only on your direct contacts, but also on contacts of contacts, etc.

The nodes clients who have been verified as infected, could have an additional field with a signature from the authorities so that they are more trusted.

Shortcomings:

1) Server wise is heavy, potentially thousand of request per second. (We can make this work, no prob.)
2) The global whiteboard becomes a single point of failure. Can be DoS, taken down, etc. Unclear the motivation but it would be possible. Mitigation, replication.
3) Suffers from some attacks, some of them shared with the proposal of PEPP. 
4) To avoid the impersonation attack, kEphID should be asymmetric, so that an attacker cannot create a fake payload just by capturing EphID and kEphID and submitting to the whiteboard. It is unclear, however, that this would fit on the broadcast bandwidth.

Apologies for the brain dump, I'd happy to attend any question you might have to the best of my abilities. And keep up the good work!

**Comment by cluck** [comment]
Note that "disregard[ing] any environmental consideration" is an intended feature of the protocol design: not recording more than necessary (proximity) reduces the risk of being later incriminated for having visited particular locations.

So, if "calculating a risk" on top of this data becomes possible, the protocol is basically flawed and no more trustworthy.

**Comment by cascremers** [comment]
Thank you! This has come up in internal design discussions, and we are considering if there are  options that might meet our requirements.

**Comment by solso** [comment]
This approach does not record anything else more than local contacts, exactly as vanilla PEPP. 

Certain location awareness becomes embedded by having knowledge about the risk of the contacts, which is calculated as a function of the risk of the contact's contacts and so on. 

Basically, you risk is the function of the risk of your contacts. This process is done iteratively (without coordination), and you should obtain on the steady state, the risk of each individuals. Given a set of seed (people infected or self-reported). Think of it as a distributed page-rank (forgive for the analogy).

The main asset of considering n-th degree contacts is a more dynamic risk assessment, as there is no need to direct contacts, but also accounts for indirect contacts (which account -- to a certain limited degree) through the environment. The indirect contacts, however, only work as a proxy for environment when there is a chain of contacts over time (basically a hub/relay on the terminology of delay tolerant networks).

Hope the concern is addressed. Thanks,

**Comment by solso** [comment]
> Thank you! This has come up in internal design discussions, and we are considering if there are options that might meet our requirements.

Let me know if it needs further elaboration, have thought about it for a while, but of course, for sure it has blind-spots have I have missed. 

On additional thing. Engineering-wise do not concern on the requirements of the whiteboard. I'm on a position to commit external resources for free on that angle.

**Comment by cluck** [comment]
This sentence
> This approach does not record anything else more than local contacts, exactly as vanilla PEPP.

does not reflect what I read in https://gist.github.com/solso/d3d298b75eaa917b855773118f0872d6.

The proposal differs fundamentally from vanilla PEPP in at least three aspects:

-  it labels close-contacts with time-stamps and records repeated contacts every time they re-occur; this differs from logging random BLE addresses once per day into daily bins (anyone gaining access to this labeled data can read much more out of it, e.g. if somebody gathered with other people for more than 15 minutes);

 - it represents close-contacts not just by exchanging random IDs, but by exchanging a _bid_ which contains a unique identifier `(unique-identifier, BL_ADDR)`. In case of infection this unique identifier is meant to be broadcast to everybody (as plaintext or as an oracle);

 - it changes the protocol from a mostly-offline and clearly unidirectional protocol, to a mostly-online and bidirectional protocol (it requires that contacts above some threshold are sent to a central server). This add the requirement for every bluetooth device to also be a mobile-connected device, and thus makes them subject to already established mass surveillance systems. It basically makes every pass by the cashier in any grocery store a risky close-contact which becomes immediately notable in the mobile network mass surveillance system, and thus easily re-correlatable.

Moreover the linked document suggests that hashing the bid with SHA256 would prevent identifiability. This does not withstand even basic cryptographic analysis. Just extorting (or mass-collecting) the unique identifier allows for breaking the hashes in the broadcast bids (the space of BLE random addresses is cryptographically weak: 48-bit). Think of declaring electronic payment terminals as high-risk contamination surfaces, and having them start collecting bids of all people nearby.

So, this proposal opens a huge can of works about cryptographic robustness, and has a high risk of adding sneaky backdoors for mass surveillance.

**Comment by solso** [comment]
> This sentence
> 
> > This approach does not record anything else more than local contacts, exactly as vanilla PEPP.
> 
> does not reflect what I read in https://gist.github.com/solso/d3d298b75eaa917b855773118f0872d6.
> 

You are right, I should have said that the approach proposed does not require anything else that the contact discovery already described in PEPP. The one described on the gist (written before PEPP was announced) is by far weaker. No argument here.

My apologies for not being precise. Also, the line is very troubled as I also wanted to address what I understood was a confusion between "environment" and "locations".

>> Note that "disregard[ing] any environmental consideration" is an intended feature of the protocol design: not recording more than necessary (proximity) reduces the risk of being later incriminated for having visited particular locations.

Relying on PEPP contacts scheme should address the first 2 points/criticisms (although your latest point about broadcasting is missleading, only those who have seen the b_id should be able to decrypt based on the key.)

Let me address what I understand is the main and only point (once cleared that this is predicated upon PEPP contact scheme).

>     * it changes the protocol from a mostly-offline and clearly unidirectional protocol, to a mostly-online and bidirectional protocol (it requires that contacts above some threshold are sent to a central server). This add the requirement for every bluetooth device to also be a mobile-
connected device, and thus makes them subject to already established mass surveillance systems. It basically makes every pass by the cashier in any grocery store a risky close-contact which becomes immediately notable in the mobile network mass surveillance system, and thus easily re-correlatable.
> 
> 

Requires contacts above some threshold are send to a central server. Not exactly. If the get operation, to recover the risk of a contact you have had is done by EphID then yes, it open the door to study the frequency but that's not how it was described. Rather, EphID are accessed by the truncated hash, first byte for instance, so for each get key you will get a set of EphID's (e.g. thousands). 

This mechanism is exactly what browsers do for phishing protection: to alert for phishing urls without disclosing which page the user is visiting. There is no mass surveillance here, provided that the groups are large enough. What there is, however, is an important data transfer overhead.

Correct, it assumes that devices to have internet access.

> Moreover the linked document suggests that hashing the bid with SHA256 would prevent identifiability. This does not withstand even basic cryptographic analysis. Just extorting (or mass-

My suggestion was that rather to use plain BL_ADDR, at least hash it, not that would make impossible a dictionary attack. 

Please, do not treat this (or any other issue) as academic papers, where one focuses on details to have an excuse to reject it. 

The underlying idea is always more valuable than the particular implementation proposed. I'm sure that someone else can come up with a better implementation.

Having access to contacts of contacts will give a much earlier warning to potentially infected people, as you do not have to wait until a direct contact of yours is declared positive. IMHO, this is something worth spending some brain cycles on.

Btw, I have absolutely no attachment to particular approach. If there is something that works better, I'm all for it. Hands down.

**Comment by diedicar** [comment]
I tried to achieve tracing of contacts of contacts with the very simple scheme presented in issue #85 and #86 .
